### ISSUE TITLE
String.replace(regexp|substr, newSubstr|function)

### ISSUE THREAD

**Issue by jasonwilliams** [boa-dev/boa#116]
`replace` function would need to be added to the string object here:
https://github.com/jasonwilliams/boa/blob/master/src/lib/js/string.rs

It would be very similar to:
https://github.com/jasonwilliams/boa/blob/master/src/lib/js/regexp.rs#L212-L251

Spec:
https://tc39.es/ecma262/#sec-string.prototype.replace

Notes:
https://github.com/jasonwilliams/boa/blob/master/docs/debugging.md

**Comment by louisgjohnson** [comment]
Hey mate, can I have a go of this?

**Comment by thiagoarrais** [comment]
@louisgjohnson: still working on this? I would like to help with it.

**Comment by cemremengu** [comment]
@jasonwilliams I think this is already added (as I understand from the dicussions)

**Comment by RageKnify** [comment]
@jasonwilliams you implemented this in #217, but as I noted in #629 , there is a bug with the current implementation:
Copied from Chrome:
```js
> let str = 'boa';
> str.replace(/.*/, "??$&??")
"??boa??"
> str.replace(".*", "??$&??")
"boa"
```
While on `boa` both would return `"??boa??"` because the string `".*"` is converted to a RegEx pattern.

Looking at the spec: https://tc39.es/ecma262/#sec-string.prototype.replace
I believe on point 2 if `searchValue` is a `RegExp` object it should give `RegExp.prototype [ @@replace ]` as `replacer`. Which would then be invoked. `@@replace` exists, but `RegExp.prototype [ @@replace ]` isn't implemented, I think the code in `String.replace` is an aproximation of it.

After #629 gets merged I'd like to look into implementing `RegExp.prototype[ @@replace ]` and making String.replace spec compliant.

**Comment by jedel1043** [comment]
Closing, as the bug that blocked this is solved. If we overlooked something we can open another issue with the specific problem.
